### ISSUE TITLE
Dynamic pipeline NoC

### ISSUE THREAD

**Issue by sertel** [Barkhausen-Institut/noc-koika#1]
Restructure and design the routers in order to be able to construct a NoC that is dynamic in the horizontal dimension but not in the vertical dimension.
For sake of simplicity, please start with a stateful NoC where each channel is represented by a single register.

Steps:
- [x] Create a static pipeline.
- [x] Integrate MetaCoq via Nix as a new input to the flake.
- [x] Use MetaCoq to create the inductive data types for the registers. 
- [x] Create functions for the routers that still need the `reg_t` type as an input. Use a `Module Type` to get this done.
- [x] Do the composition of a NoC dynamically.

**Comment by garvitchhabra-7** [comment]
The pipeline composition is completely dynamic now.

**Comment by garvitchhabra-7** [comment]
MetaCoq runs into some issues using Parameter when we try to generate Coq code from syntax, I'm checking if I can find a way to fix this. 

> Not Supported raised at unquote_list

**Comment by sertel** [comment]
So you mean the `Parameter` command?

**Comment by garvitchhabra-7** [comment]
Yes if I define nocsize using Parameter, the MetaCoq commands (both the command which is used to build inductive types and the command that is used to build definitions) error out, they work fine when using Definition.

**Comment by sertel** [comment]
Ah, I see.
Thanks for clarifying!
I'm looking forward to your ideas.
